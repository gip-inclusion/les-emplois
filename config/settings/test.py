import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
