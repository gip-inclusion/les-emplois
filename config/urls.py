from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path, register_converter
from django.views.generic import TemplateView

from itou.utils.urls import SiretConverter
from itou.www.dashboard import views as dashboard_views
from itou.www.login import views as login_views
from itou.www.signup import views as signup_views


register_converter(SiretConverter, "siret")

urlpatterns = [
    path("admin/", admin.site.urls),
    # --------------------------------------------------------------------------------------
    # allauth URLs. Order is important because some URLs are overriden.
    # --------------------------------------------------------------------------------------
    # Override allauth `account_signup` URL.
    # /accounts/signup/ <=> account_signup
    # We don't want any user to be able to signup using the default allauth `signup` url
    # because we have multiple specific signup processes for different kind of users.
    re_path(r"^accounts/signup/$", signup_views.signup),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_login` URL.
    # /accounts/login/ <=> account_login
    # We override this view because the login page should look slightly differently
    # for job seekers, prescribers and employers.
    # Also, PEAMU is only available for job seekers.
    re_path(r"^accounts/login/$", login_views.login),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_change_password` URL.
    # /accounts/password/change/ <=> account_change_password
    # https://github.com/pennersr/django-allauth/issues/468
    re_path(r"^accounts/password/change/$", dashboard_views.password_change),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_reset_password` URL.
    # Avoid user enumeration via password reset page.
    re_path(r"^accounts/password/reset/$", signup_views.ItouPasswordResetView.as_view()),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_logout` URL.
    # /accounts/logout/ <=> account_logout
    # We need custom code to process PEAMU logout.
    re_path(r"^accounts/logout/$", dashboard_views.logout),
    # --------------------------------------------------------------------------------------
    # Other allauth URLs.
    path("accounts/", include("allauth.urls")),
    # --------------------------------------------------------------------------------------
    # PEAMU URLs.
    path("accounts/", include("itou.allauth_adapters.peamu.urls")),
    # FranceConnect URLs.
    path("franceconnect/", include("itou.france_connect.urls")),
    # --------------------------------------------------------------------------------------
    # API.
    path("api/v1/", include("itou.api.urls", namespace="v1")),
    path("api/", include("itou.www.api.urls")),
    # www.
    path("", include("itou.www.home.urls")),
    path("apply/", include("itou.www.apply.urls")),
    path("approvals/", include("itou.www.approvals_views.urls")),
    path("autocomplete/", include("itou.www.autocomplete.urls")),
    path("dashboard/", include("itou.www.dashboard.urls")),
    path("employee_record/", include("itou.www.employee_record_views.urls")),
    path("institutions/", include("itou.www.institutions_views.urls")),
    path("invitations/", include("itou.www.invitations_views.urls")),
    path("prescribers/", include("itou.www.prescribers_views.urls")),
    path("search/", include("itou.www.search.urls")),
    path("siae/", include("itou.www.siaes_views.urls")),
    path("signup/", include("itou.www.signup.urls")),
    path("stats/", include("itou.www.stats.urls")),
    path("versions/", include("itou.www.releases.urls")),
    # Enable Anymailâ€™s status tracking
    # https://anymail.readthedocs.io/en/stable/esps/mailjet/#status-tracking-webhooks
    re_path(r"^webhooks/anymail/", include("anymail.urls")),
    path("welcoming_tour/", include("itou.www.welcoming_tour.urls")),
    # Static pages.
    path("accessibility/", TemplateView.as_view(template_name="static/accessibility.html"), name="accessibility"),
    # Errors pages.
    path("403/", TemplateView.as_view(template_name="403.html"), name="403"),
    path("404/", TemplateView.as_view(template_name="404.html"), name="404"),
    path("500/", TemplateView.as_view(template_name="500.html"), name="500"),
]

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    # In the local developpemnt environment, we enable the default callback URLs
    urlpatterns = [path("", include("itou.france_connect.urls"))] + urlpatterns
