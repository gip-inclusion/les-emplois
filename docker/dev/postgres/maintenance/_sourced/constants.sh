#!/usr/bin/env bash

export BACKUP_DIR_PATH='/backups'
export BACKUP_FILE_PREFIX='backup'
