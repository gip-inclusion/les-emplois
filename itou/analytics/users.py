from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
