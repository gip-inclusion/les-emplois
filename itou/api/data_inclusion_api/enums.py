import enum


class StructureTypeStr(str, enum.Enum):
    ORGA = "orga"
    SIAE = "siae"
