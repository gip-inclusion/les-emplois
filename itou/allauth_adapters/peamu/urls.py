from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
