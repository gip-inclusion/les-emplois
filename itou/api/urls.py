from django.urls import include, path
from drf_spectacular.views import SpectacularAPIView, SpectacularRedocView
from rest_framework import routers
from rest_framework.authtoken import views as auth_views

from itou.api.data_inclusion_api.views import DataInclusionStructureView

from .applicants_api.views import ApplicantsView
from .employee_record_api.viewsets import (
    DummyEmployeeRecordViewSet,
    EmployeeRecordUpdateNotificationViewSet,
    EmployeeRecordViewSet,
)
from .siae_api.viewsets import SiaeViewSet


# High level app for API
# https://docs.djangoproject.com/en/dev/topics/http/urls/#url-namespaces-and-included-urlconfs
app_name = "itou.api"

# Using DRF router with viewsets means automatic definition of utl patterns
router = routers.DefaultRouter()
router.register(r"employee-records", EmployeeRecordViewSet, basename="employee-records")
router.register(
    r"employee-record-notifications",
    EmployeeRecordUpdateNotificationViewSet,
    basename="employee-record-notifications",
)
router.register(r"dummy-employee-records", DummyEmployeeRecordViewSet, basename="dummy-employee-records")
router.register(r"siaes", SiaeViewSet, basename="siaes")

urlpatterns = [
    # TokenAuthentication endpoint to get token from login/password.
    path("token-auth/", auth_views.obtain_auth_token, name="token-auth"),
    # Needed for Browseable API (dev)
    path("api-auth/", include("rest_framework.urls", namespace="rest_framework")),
    # OpenAPI
    # See: https://www.django-rest-framework.org/topics/documenting-your-api/
    # OAS 3 YAML schema (downloadable)
    path(
        "oas3/",
        SpectacularAPIView.as_view(),
        name="openapi_schema",
    ),
    path(
        "redoc/",
        SpectacularRedocView.as_view(url_name="v1:openapi_schema"),
        name="redoc",
    ),
]

urlpatterns += router.urls

urlpatterns += [
    path("structures/", DataInclusionStructureView.as_view(), name="structures-list"),
    path("candidats/", ApplicantsView.as_view(), name="applicants-list"),
]
