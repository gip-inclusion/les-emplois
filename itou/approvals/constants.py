from itou.approvals.enums import ProlongationReason


# A prolongation report file can be uploaded only for these reasons
PROLONGATION_REPORT_FILE_REASONS = (
    ProlongationReason.RQTH,
    ProlongationReason.SENIOR,
    ProlongationReason.PARTICULAR_DIFFICULTIES,
)
