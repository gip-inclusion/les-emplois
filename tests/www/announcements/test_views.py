import pytest
from django.urls import reverse
from pytest_django.asserts import assertContains, assertRedirects

from itou.users.enums import UserKind
from itou.utils.constants import ITOU_HELP_CENTER_URL
from tests.communications.factories import AnnouncementCampaignFactory, AnnouncementItemFactory
from tests.users.factories import EmployerFactory, JobSeekerFactory, PrescriberFactory
from tests.utils.test import parse_response_to_soup


class TestNewsRender:
    @pytest.fixture(autouse=True)
    def empty_announcements_cache(self, empty_active_announcements_cache):
        pass

    def _assert_all_items_rendered(self, response, *campaigns):
        assert response.status_code == 200

        for i, rendered_campaign in enumerate(response.context["news_page"]):
            rendered_items = rendered_campaign.items.all()
            assert list(rendered_items) == list(campaigns[i].items.all())

    def test_all_news_rendered_html(self, client, snapshot):
        campaign = AnnouncementCampaignFactory(for_snapshot=True, with_items_for_every_user_kind=True)

        user = EmployerFactory(with_company=True)
        client.force_login(user)
        response = client.get(reverse("announcements:news"))

        self._assert_all_items_rendered(response, campaign)

        # rendering in HTML
        content = parse_response_to_soup(response, ".s-section__container")
        assert str(content) == snapshot

    def test_announcements_anonymous_user(self, client):
        campaign = AnnouncementCampaignFactory(for_snapshot=True, with_items_for_every_user_kind=True)

        response = client.get(reverse("announcements:news"))
        self._assert_all_items_rendered(response, campaign)

    def test_campaign_items_filtered_on_user_type(self, client):
        campaign = AnnouncementCampaignFactory(with_items_for_every_user_kind=True)
        url = reverse("announcements:news")

        # prescriber receives all items
        user = PrescriberFactory()
        client.force_login(user)
        response = client.get(url)
        self._assert_all_items_rendered(response, campaign)

        # candidates only receive news relating to them
        assert campaign.items.exclude(user_kind_tags__contains=[UserKind.JOB_SEEKER]).exists()
        client.force_login(JobSeekerFactory())
        response = client.get(url)
        assert response.status_code == 200

        items = response.context["news_page"][0].items
        for tags in items.values_list("user_kind_tags", flat=True):
            assert len(tags) == 0 or UserKind.JOB_SEEKER.value in tags
        assert items.count() < campaign.items.count()
        assert items.count() == response.context["news_page"][0].count_items

    def test_only_live_campaigns_rendered(self, client):
        campaign = AnnouncementCampaignFactory(with_item=True)
        AnnouncementCampaignFactory(with_item=True, live=False)
        AnnouncementCampaignFactory(live=True)  # no item

        response = client.get(reverse("announcements:news"))
        assert response.status_code == 200
        assert len(response.context["news_page"]) == 1
        assert response.context["news_page"][0] == campaign

    def test_pagination(self, client):
        items_per_page = 12
        total_items = items_per_page + 2
        campaigns = AnnouncementCampaignFactory.create_batch(total_items, with_items_for_every_user_kind=True)

        url = reverse("announcements:news")
        second_page_url = f"{url}?page=2"

        response = client.get(url)
        assertContains(response, second_page_url)
        assert len(response.context["news_page"]) == items_per_page

        response = client.get(second_page_url)
        assert response.status_code == 200
        expected_len = total_items - items_per_page
        assert len(response.context["news_page"]) == expected_len
        paged_campaigns = campaigns[items_per_page : (items_per_page + expected_len)]
        self._assert_all_items_rendered(response, *paged_campaigns)

    def test_none_exists(self, client):
        expected_redirect_url = f"{ ITOU_HELP_CENTER_URL }/categories/25225629682321--Nouveaut%C3%A9s"

        client.force_login(PrescriberFactory())
        url = reverse("announcements:news")
        assertRedirects(client.get(url), expected_redirect_url, fetch_redirect_response=False)

        AnnouncementCampaignFactory(live=False, with_item=True)
        assertRedirects(client.get(url), expected_redirect_url, fetch_redirect_response=False)

        # it's also possible that there are live campaigns without content for my user kind
        AnnouncementItemFactory(user_kind_tags=[UserKind.PRESCRIBER])
        client.force_login(JobSeekerFactory())
        assertRedirects(client.get(url), expected_redirect_url, fetch_redirect_response=False)
