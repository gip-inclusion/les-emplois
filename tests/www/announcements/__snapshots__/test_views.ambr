# serializer version: 1
# name: TestNewsRender.test_all_news_rendered_html
  '''
  <div class="s-section__container container">
              <ul class="list-group list-group-collapse list-group-flush">
                  
                      <li class="list-group-item list-group-item-action">
                          <button aria-controls="news-1" aria-expanded="true" class="d-block w-100" data-bs-target="#news-1" data-bs-toggle="collapse" type="button">
                              <div class="h3 mb-0">
                                  Janvier 2024
                                  <span class="badge badge-sm bg-emploi-light text-info rounded-pill ms-2">6
                                      <span class="visually-hidden">articles</span>
                                  </span>
                              </div>
                          </button>
                          <article aria-controls="news-1" class="mt-3 collapse show" id="news-1">
                              
                                  <div class="row mb-3">
                                      <div class="col-12 col-md-4 d-none d-md-inline">
                                          
                                              
                                              
                                              <img alt="" aria-hidden="true" class="img-fitcover img-thumbnail img-muted" height="396" loading="lazy" src="/static/vendor/theme-inclusion/images/ico-bicro-important.svg" width="420"/>
                                          
                                      </div>
                                      <div class="col-12 col-md-8">
                                          
                                          <p class="h4 my-2">Nouvelle fonctionnalité sur notre site</p>
                                          <p>Il est désormais possible de faire de nouvelles actions avec votre compte</p>
                                          
                                              <a aria-label="Plus d'informations sur Nouvelle fonctionnalité sur notre site" class="btn-link has-external-link" href="https://example.com/" rel="noopener" target="_blank">En savoir plus</a>
                                          
                                      </div>
                                  </div>
                              
                                  <div class="row mb-3">
                                      <div class="col-12 col-md-4 d-none d-md-inline">
                                          
                                              
                                              
                                              <img alt="" aria-hidden="true" class="img-fitcover img-thumbnail img-muted" height="396" loading="lazy" src="/static/vendor/theme-inclusion/images/ico-bicro-important.svg" width="420"/>
                                          
                                      </div>
                                      <div class="col-12 col-md-8">
                                          
                                              <div aria-label="groupes d'utilisateurs concernés">
                                                  
                                                      <span class="tag tag-base bg-info-lighter text-info">CANDIDAT</span>
                                                  
                                              </div>
                                          
                                          <p class="h4 my-2">Nouvelle fonctionnalité sur notre site</p>
                                          <p>Il est désormais possible de faire de nouvelles actions avec votre compte</p>
                                          
                                              <a aria-label="Plus d'informations sur Nouvelle fonctionnalité sur notre site" class="btn-link has-external-link" href="https://example.com/" rel="noopener" target="_blank">En savoir plus</a>
                                          
                                      </div>
                                  </div>
                              
                                  <div class="row mb-3">
                                      <div class="col-12 col-md-4 d-none d-md-inline">
                                          
                                              
                                              
                                              <img alt="" aria-hidden="true" class="img-fitcover img-thumbnail img-muted" height="396" loading="lazy" src="/static/vendor/theme-inclusion/images/ico-bicro-important.svg" width="420"/>
                                          
                                      </div>
                                      <div class="col-12 col-md-8">
                                          
                                              <div aria-label="groupes d'utilisateurs concernés">
                                                  
                                                      <span class="tag tag-base bg-info-lighter text-info">PRESCRIPTEUR</span>
                                                  
                                              </div>
                                          
                                          <p class="h4 my-2">Nouvelle fonctionnalité sur notre site</p>
                                          <p>Il est désormais possible de faire de nouvelles actions avec votre compte</p>
                                          
                                              <a aria-label="Plus d'informations sur Nouvelle fonctionnalité sur notre site" class="btn-link has-external-link" href="https://example.com/" rel="noopener" target="_blank">En savoir plus</a>
                                          
                                      </div>
                                  </div>
                              
                                  <div class="row mb-3">
                                      <div class="col-12 col-md-4 d-none d-md-inline">
                                          
                                              
                                              
                                              <img alt="" aria-hidden="true" class="img-fitcover img-thumbnail img-muted" height="396" loading="lazy" src="/static/vendor/theme-inclusion/images/ico-bicro-important.svg" width="420"/>
                                          
                                      </div>
                                      <div class="col-12 col-md-8">
                                          
                                              <div aria-label="groupes d'utilisateurs concernés">
                                                  
                                                      <span class="tag tag-base bg-info-lighter text-info">EMPLOYEUR</span>
                                                  
                                              </div>
                                          
                                          <p class="h4 my-2">Nouvelle fonctionnalité sur notre site</p>
                                          <p>Il est désormais possible de faire de nouvelles actions avec votre compte</p>
                                          
                                              <a aria-label="Plus d'informations sur Nouvelle fonctionnalité sur notre site" class="btn-link has-external-link" href="https://example.com/" rel="noopener" target="_blank">En savoir plus</a>
                                          
                                      </div>
                                  </div>
                              
                                  <div class="row mb-3">
                                      <div class="col-12 col-md-4 d-none d-md-inline">
                                          
                                              
                                              
                                              <img alt="" aria-hidden="true" class="img-fitcover img-thumbnail img-muted" height="396" loading="lazy" src="/static/vendor/theme-inclusion/images/ico-bicro-important.svg" width="420"/>
                                          
                                      </div>
                                      <div class="col-12 col-md-8">
                                          
                                              <div aria-label="groupes d'utilisateurs concernés">
                                                  
                                                      <span class="tag tag-base bg-info-lighter text-info">INSPECTEUR DU TRAVAIL</span>
                                                  
                                              </div>
                                          
                                          <p class="h4 my-2">Nouvelle fonctionnalité sur notre site</p>
                                          <p>Il est désormais possible de faire de nouvelles actions avec votre compte</p>
                                          
                                              <a aria-label="Plus d'informations sur Nouvelle fonctionnalité sur notre site" class="btn-link has-external-link" href="https://example.com/" rel="noopener" target="_blank">En savoir plus</a>
                                          
                                      </div>
                                  </div>
                              
                                  <div class="row mb-3">
                                      <div class="col-12 col-md-4 d-none d-md-inline">
                                          
                                              
                                              
                                              <img alt="" aria-hidden="true" class="img-fitcover img-thumbnail img-muted" height="396" loading="lazy" src="/static/vendor/theme-inclusion/images/ico-bicro-important.svg" width="420"/>
                                          
                                      </div>
                                      <div class="col-12 col-md-8">
                                          
                                              <div aria-label="groupes d'utilisateurs concernés">
                                                  
                                                      <span class="tag tag-base bg-info-lighter text-info">ADMINISTRATEUR</span>
                                                  
                                              </div>
                                          
                                          <p class="h4 my-2">Nouvelle fonctionnalité sur notre site</p>
                                          <p>Il est désormais possible de faire de nouvelles actions avec votre compte</p>
                                          
                                              <a aria-label="Plus d'informations sur Nouvelle fonctionnalité sur notre site" class="btn-link has-external-link" href="https://example.com/" rel="noopener" target="_blank">En savoir plus</a>
                                          
                                      </div>
                                  </div>
                              
                          </article>
                      </li>
                  
                  
              </ul>
          </div>
  '''
# ---
# name: TestNewsRender.test_none_exists[none_exists]
  '''
  <div class="s-section__container container">
              <ul class="list-group list-group-collapse list-group-flush">
                  
                  <p>Aucune nouveauté pour le moment.</p>
              </ul>
          </div>
  '''
# ---
